google/gax/api_callable.py

```diff
@@ -306,8 +293,64 @@ def _upper_camel_to_lower_under(string):
     return out
 
 
+def _override_retry(config_base, overrides):
+    """Update config_base specifying retry params with overrides.
+
+    Args:
+      config_base: A dictionary of retry params.
+      overrides: A dictionary to update config_base.
+    """
+    for (name, params) in overrides.items():
+        if name in config_base:
+            config_base[name].update(params)
+        else:
+            config_base[name] = params
+
+
+def _override_methods(config_base, overrides):
+    """Update config_base specifying method condigs with overrides.
+
+    Args:
+      config_base: A dictionary of method config.
+      overrides: A dicitionary to update config_base.
+    """
+    for (name, params) in overrides.items():
+        if not params:
+            config_base[name] = None
+        else:
+            for (key, param) in params.items():
+                if key not in config_base[name]:
+                    continue
+                if key == 'bundling' and param:
+                    config_base[name][key].update(param)
+                else:
+                    config_base[name][key] = param
+
+
+def _override_config(config_base, overrides):
+    """Merge overriding configs into config_base and return the new config.
+
+    If overrides is empty, it returns the same config_base. Otherwise,
+    it creates a copy of config_base and some parts are updated by overrides.
+
+    Args:
+      config_base: A dictionary for base config.
+      overrides: A dictionary in the same structure of config_base.
+    """
+    if not overrides:
+        return config_base
+    new_config = copy.deepcopy(config_base)
+    if 'retry_codes' in overrides:
```

Comment:
I believe this is what we want.

Here, we want to allow users to provide a copy (with edits) of JSON config. Users can copy the JSON file, modify it as they want, and then pass it as a parameter. This is why the new override has the exactly same structure as the original config JSON.

That means, anything specified by the users has to supersede the default config. Otherwise, some of edits in the config will be silently ignored and the default values are used -- that's a more surprising behavior.